blockchain/indexers/utreexocfindex.go

```diff
@@ -0,0 +1,370 @@
+// interface.
+func (idx *UtreexoCFIndex) Init(chain *blockchain.BlockChain) error {
+	idx.chain = chain
+	return nil // Nothing to do.
+}
+
+// Key returns the database key to use for the index as a byte slice. This is
+// part of the Indexer interface.
+func (idx *UtreexoCFIndex) Key() []byte {
+	return utreexoCFIndexParentBucketKey
+}
+
+// Name returns the human-readable name of the index. This is part of the
+// Indexer interface.
+func (idx *UtreexoCFIndex) Name() string {
+	return utreexoCFIndexName
+}
+
+// Create is invoked when the index manager determines the index needs to
+// be created for the first time. It creates buckets for the custom utreexo
+// filter index.
+func (idx *UtreexoCFIndex) Create(dbTx database.Tx) error {
+	meta := dbTx.Metadata()
+
+	utreexoCfIndexParentBucket, err := meta.CreateBucket(utreexoCFIndexParentBucketKey)
+	if err != nil {
+		return err
+	}
+
+	for _, bucketName := range utreexoCfHeaderKeys {
+		_, err = utreexoCfIndexParentBucket.CreateBucket(bucketName)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// storeUtreexoCFilter stores a given utreexocfilter header
+func (idx *UtreexoCFIndex) storeUtreexoCFHeader(dbTx database.Tx, block *btcutil.Block, filterData []byte,
+	filterType wire.FilterType) error {
+	if filterType != wire.UtreexoCFilter {
+		return errors.New("invalid filter type")
+	}
+
+	// Figure out which header bucket to use.
+	hkey := utreexoCfHeaderKeys[0]
+	h := block.Hash()
+
+	// fetch the previous block's filter header.
+	var prevHeader *chainhash.Hash
+	ph := &block.MsgBlock().Header.PrevBlock
+	if ph.IsEqual(&zeroHash) {
+		prevHeader = &zeroHash
+	} else {
+		var pfh []byte
+		err := idx.db.View(func(dbTx database.Tx) error {
+			var err error
+			pfh, err = dbFetchUtreexoCFilterIdxEntry(dbTx, hkey, ph)
+			return err
+		})
+		if err != nil {
+			return err
+		}
+
+		// Construct the new block's filter header, and store it.
+		prevHeader, err = chainhash.NewHash(pfh)
+		if err != nil {
+			return err
+		}
+	}
+
+	fh, err := builder.MakeHeaderForUtreexoCFilter(filterData, *prevHeader)
+	if err != nil {
+		return err
+	}
+	return dbStoreUtreexoCFilterIdxEntry(dbTx, hkey, h, fh[:])
+}
+
+// ConnectBlock is invoked by the index manager when a new block has been
+// connected to the main chain.
+// This is part of the Indexer interface.
+func (idx *UtreexoCFIndex) ConnectBlock(dbTx database.Tx, block *btcutil.Block,
+	stxos []blockchain.SpentTxOut) error {
+
+	blockHash := block.Hash()
+	roots, leaves, err := idx.fetchUtreexoRoots(blockHash, block)
+
+	if err != nil {
+		return err
+	}
+
+	// serialize the hashes of the utreexo roots hash
+	serializedUtreexo, err := blockchain.SerializeUtreexoRootsHash(leaves, roots)
+	if err != nil {
+		return err
+	}
+
+	return idx.storeUtreexoCFHeader(dbTx, block, serializedUtreexo, wire.UtreexoCFilter)
+}
+
+// fetches the utreexo roots for a given block hash
+func (idx *UtreexoCFIndex) fetchUtreexoRoots(blockHash *chainhash.Hash,
```

Comment:
I don't having a single `fetchUtreexoRoots()` works because you have 3 different scenarios.

1: Fetching roots when the index is caught up.
2: Fetching when the index isn't caught up. This happens when you enable an index after syncing the chain first.
3: Fetching for when peers ask for the roots.

For 1, you should always call `FetchCurrentUtreexoState()`.

For 2, you wanna call `FetchUtreexoState()` or `FetchCurrentUtreexoState()` depending on whether or not the index is all caught up or not.

For 3, you want to copy how the fetching is done in `rpcserver.go`https://github.com/utreexo/utreexod/blob/main/rpcserver.go#L3529-L3530

Reply:
So that'll mean i don't have to copy the implementation of how roots fetching from rpcserver.go works, right?
Because I initially had that code, but i totally overlooked one possibility you mentioned above, so when trying to catch indexes with that part of the code included, it throws an error.